syv followup constituency first

baob opened the pull request:
This will close #625 and possibly #922 too.

Most of the work done to move the constituency-first commits over, making constituency the first question asked during registration.

The API still works.

User edit (constituency) confirmed working

baob committed:
split home page view into partials

baob committed:
change the flow to constituency first

This is a combination of 7 commits.

HACKING towards constituency first - step 1-6

unsuccessful in getting the hidden field really hidden. I suspect some CSS skills required

baob committed:
specs updated and passing

aspiers committed:
fix linting

aspiers committed:
lower minimum branch coverage

baob committed:
fix 3 tests, now the home page lets you choose constituency, not candidate

baob committed:
COMMENT OUT FAILING TESTS FOR NOW

the spec should be moved to the page that does the constituency selection

baob committed:
fix the failing test by ensuring there is always a constituency in the database

baob committed:
fix the no method error for nil on parties

Relates to https://github.com/swapmyvote/swapyourvote-usa/issues/24

1. the method prepopulate_fields_from_session was referencing parties
   before the state had even been defined (part of the api/prepopulate
   code)

2. the 'pre_login' code that sequences state-then-candidates questions
   was unecessarily messing with the session parameters associated with
   api/prepopulate.

baob committed:
add postcode field to constituency selection

This is a combination of 3 commits.

add postcode field to constituency selection BUT NOT WORKING
add postcodes helper to home page (now has constituencies form)
fixed postcode helper to work on home page

baob committed:
update snapshots postcode field

baob committed:
fix font size for constituency form on home page

baob committed:
hide the constituency auto-complete on the candidate selection page

baob committed:
add pre-population specs for pre-login page

partially addresses comment on commit 3a94e853

baob committed:
fix pre-population specs for home page

now fully addresses comment on commit 3a94e853

baob commented:
> Some files could not be reviewed due to errors:

Raised issue #1023 to address this

baob committed:
better text for home page CTA

stevebaxter commented on app/views/home/_constituency_form.html.haml:
```diff
@@ -0,0 +1,14 @@
+= form_with url: pre_login_path, id: "login-form", local: true do |f|
```
So I think here you want to do:

```
.d-flex.flex-column.justify-content-center
```

See here for example:

https://github.com/swapmyvote/swapmyvote/blob/8ca61f9a1ba59e87ee61d4e50ad55a1f28339ecf/app/views/user/swaps/_swap_profile.html.haml#L1

This should then lay out all the divs vertically and centred horizontally. You can read about the built-in BS flex classes here:

https://getbootstrap.com/docs/5.3/utilities/flex/

stevebaxter replied:
Sorry, try this:
```
= form_with url: pre_login_path, id: "login-form", local: true do |f|

  .d-flex.flex-column.align-items-center

    %p.choose-constituency
      My constituency is
      = f.collection_select :constituency_ons_id, constituencies,
                            :ons_id, :name,
                            selected: default_constituency_ons_id,
                            prompt: true

    = render partial: "users/postcode_field"

    %p.text-center
      = submit_tag("Next: Choose Parties", class: "btn btn-primary")

```
The spacing still isn't right because the CSS is a bit inconsistent, this will centre it though.

stevebaxter commented on app/views/home/_constituency_form.html.haml:
```diff
@@ -0,0 +1,14 @@
+= form_with url: pre_login_path, id: "login-form", local: true do |f|
+
+
+  %p.choose-constituency
```
I'm not sure you want to use paragraphs for this, instead I would use divs with standard padding so each div gets its own padding. At the moment it isn't quite right, as the postcode field doesn't get the right padding.

stevebaxter commented on app/views/home/_constituency_form.html.haml:
```diff
@@ -0,0 +1,14 @@
+                          selected: default_constituency_ons_id,
+                          prompt: true
+
+  = render partial: "users/postcode_field"
```
You might need to put this in a div to make sure it is rendered with the right padding.

baob committed:
center-align postode sub-form

baob committed:
center-align postode sub-form - fixed snapshots

baob commented:
# Preview 

![image](https://github.com/user-attachments/assets/bc46285f-f053-4795-81cc-3812700626df)


![image](https://github.com/user-attachments/assets/1e750f0e-9abf-4613-ade0-3ee1d4082805)

stevebaxter commented on app/views/users/_postcode_field.html.haml:
```diff
@@ -1,11 +1,12 @@
-                                      maxlength: 9, spellcheck: "false" }
-    .input-group-append
-      %input.btn.btn-secondary#btn-postcode{ type: "button", value: "Search" }
+.d-flex.flex-row.justify-content-center
```
I don't think this is right. We don't want to always centre the postcode field. Instead I think you need to do what I suggested here?

https://github.com/swapmyvote/swapmyvote/pull/1024/files#r1835562684

baob committed:
Revert "center-align postode sub-form"

This reverts commit ceeb58ecf8e72102be0d8cfedfb77ca9eec3f2c0.

Addresses comment from @stevebaxter here

https://github.com/swapmyvote/swapmyvote/pull/1024/files/02f453739b2f0c5b2e2815c29d330cc1464f1034#r1835805752

baob committed:
centre-align postcode sub-form ONLY on home page

Follows @stevebaxter suggestion here

https://github.com/swapmyvote/swapmyvote/pull/1024/files#r1835562684

baob committed:
fix snapshots for alignment change

baob committed:
revert postcodesHelper so it works on user edit page

baob committed:
make home page form based on user, so postcodeHelper is re-usable on both forms

baob committed:
fix specs and snapshots for form changes

baob committed:
issue-1023: reduce local rubocop version to highest one compatible with hound CI

baob committed:
correct rubocop complaint

baob committed:
pass explicit parameter to _constituency_form rather than relying on globals

baob commented:
> Some files could not be reviewed due to errors:
> 
> Error: unrecognized cop or department Rails/I18nLocaleTexts found in .rubocop...
> Error: unrecognized cop or department Rails/I18nLocaleTexts found in .rubocop.yml
> Did you mean `Rails/I18nLocaleAssignment`?

This is really not straightforward to fix, certainly not as part of this PR. See https://github.com/swapmyvote/swapmyvote/issues/1023

tdg commented:
These look great. 

The change of call to action is radical but probably overdue. 

Can I suggest the second screen have as its button whatever the next step is, in the new onboarding journey, for example: 
"Next: find my swap"
or
"Next: verify"

baob committed:
updated CTA on pre-login choose-party page - now says: Next: Sign Up

baob commented:
@tdg The sequence is now

<img width="783" alt="image" src="https://github.com/user-attachments/assets/9f1f04a4-ea53-4468-b458-6f754e48d9de">

<img width="963" alt="image" src="https://github.com/user-attachments/assets/5f757217-db65-4e16-8ca7-a2f594cb4df1">

<img width="519" alt="image" src="https://github.com/user-attachments/assets/616433f7-cfef-475c-87c1-c8152ddd116d">

tdg commented:
Nice, sign up is great

tdg commented:
Maybe the last one should say "confirm and see swaps"

baob committed:
fix error: First argument in form cannot contain nil or be empty

baob committed:
change CTA text on sign up page (tom suggestion)
